### ISSUE TITLE
Any implementation of LRN?

### ISSUE THREAD

**Issue by shiyemin** [fchollet/keras#1549]
I am new to Keras. I just find that LRN2D has been removed and i can not find any related code in Keras. This stop me converting GoogLenet from Caffe to Keras. So, is there any hint for doing this conversion?

I have got a LRN implementation like this:

class LRN2D(Layer):
    """
    This code is adapted from pylearn2.
    License at: https://github.com/lisa-lab/pylearn2/blob/master/LICENSE.txt
    """

```
def __init__(self, alpha=1e-4, k=2, beta=0.75, n=5, **kwargs):
    if n % 2 == 0:
        raise NotImplementedError("LRN2D only works with odd n. n provided: " + str(n))
    super(LRN2D, self).__init__(**kwargs)
    self.alpha = alpha
    self.k = k
    self.beta = beta
    self.n = n

def get_output(self, train):
    X = self.get_input(train)
    b, ch, r, c = K.shape(X)
    half_n = self.n // 2
    input_sqr = K.square(X)
    extra_channels = K.zeros((b, ch + 2 * half_n, r, c))
    input_sqr = K.concatenate([extra_channels[:, :half_n, :, :],
                               input_sqr,
                               extra_channels[:, half_n + ch:, :, :]],
                              axis=1)
    scale = self.k
    for i in range(self.n):
        scale += self.alpha * input_sqr[:, i:i + ch, :, :]
    scale = scale ** self.beta
    return X / scale

def get_config(self):
    config = {"name": self.__class__.__name__,
              "alpha": self.alpha,
              "k": self.k,
              "beta": self.beta,
              "n": self.n}
    base_config = super(LRN2D, self).get_config()
    return dict(list(base_config.items()) + list(config.items()))
```

However, the "extra_channels = K.zeros((b, ch + 2 \* half_n, r, c))" is error.  
Is there anyone can help with this?

**Comment by keunwoochoi** [comment]
FYI, it's the same as [LRN2d we had in Keras](https://github.com/fchollet/keras/commit/82353da4dc66bc702a74c6c233f3e16b7682f9e6) and it worked for me.

**Comment by shiyemin** [comment]
@keunwoochoi Hi, I failed when compile with "model.compile(loss={'prob': 'mse'}, optimizer='sgd')". 

The error msg is:
![fail](https://cloud.githubusercontent.com/assets/1501158/12569965/b3914ffc-c40e-11e5-9ad6-9412ad7ea61a.png)

Do you have any ideas? 

By the way, the batch_size is not set and variable 'b' should be NONE at this position. Is this normal?

**Comment by keunwoochoi** [comment]
Hi, I'm sorry but I haven't looked into the code so far. Probably the core layer has been changed since it was removed.

**Comment by fchollet** [comment]
It would work if the shape of your input were fully defined, which it isn't (causing a bug only with TensorFlow, which you are using).

The solution is to define the shape of your input via the `batch_input_size` keyword argument.

**Comment by shiyemin** [comment]
@keunwoochoi @fchollet Thanks a lot. I fix this after using batch_input_shape and modify the caffe_to_keras code. 

Anyone who need GoogLenet may look at https://github.com/shiyemin/keras/commit/02d1cf6b575db0bdec71a2d54030b4396a27bb67.

**Comment by lazydroid** [comment]
@shiyemin do you by any chance have a working (converted?) GoogLeNet in Keras? I'd be glad to take a look at one, but I'm not sure I'll be able to do the full conversion from the very beginning.

**Comment by shiyemin** [comment]
@lazydroid Sorry, i finally give up on Keras because of its over-packaging, which makes it hard to implement my network structure. I implement my model based on https://github.com/ethereon/caffe-tensorflow.

**Comment by bhack** [comment]
Any news on this?

**Comment by keunwoochoi** [comment]
LRN is gone. A quick and dirty workaround would be reshape-BN-reshape_back. Not sure how much it would be memory-demanding.

**Comment by bhack** [comment]
It could be still useful for pretrained conversion compatiblity now that it will be soon in sync agian with Keras master https://github.com/MarcBS/keras/issues/7

**Comment by ckoren1975** [comment]
I actually got Googlenet to work with Tensorflow backend by taking a few advises I found above.  I converted LRN from theano implementation to TF.   For some reason my accuracy with the weights: https://drive.google.com/open?id=0B319laiAPjU3RE1maU9MMlh2dnc
is only 83% on top5 instead of 90% I expected.

**Comment by buckleytoby** [comment]
@ckoren1975 do you have your code for the custom LRN layer? Thanks!

**Comment by ckoren1975** [comment]
Sure, find it here:
https://github.com/ckoren1975/

under googlenet_custom_layers.py

**Comment by KissyZhou** [comment]
@ckoren1975 i think [this lrn implementation](https://github.com/ckoren1975/Machine-learning/blob/master/googlenet_custom_layers.py) is not correct.
ref: [caffe's lrn introduction](http://caffe.berkeleyvision.org/tutorial/layers/lrn.html)

problem in feature map 128x28x28x96: 
1.sum over local_size channels that's to say we should sum over 1x1xlocal_size(ie: 1x1x5) to get the final scale, not 1x1xchannels(ie: 1x1x96, in this situation we were sum over local_size=96!?)
2.there is no 'sum over channels' operation...

if i am wrong, please tell me. thanks

**Comment by sven-mayer** [comment]
In case you use the code anyway, beware there is a bug in the code, which you will hit when you load the model. Thus, remove `"name": self.__class__.__name__,`

For further details, see https://github.com/keras-team/keras/issues/11210
